crates/update-notifier/src/ui.rs

```diff
@@ -0,0 +1,66 @@
+use terminal_size::{terminal_size, Width};
+
+use crate::UpdateNotifierError;
+pub mod utils;
+
+const DEFAULT_PADDING: usize = 8;
+
+pub fn message(text: &str) -> Result<(), UpdateNotifierError> {
+    let size = terminal_size();
+    let lines: Vec<&str> = text.split('\n').map(|line| line.trim()).collect();
+
+    // get the display width of each line so we can center it within the box later
+    let lines_display_width: Result<Vec<usize>, utils::GetDisplayLengthError> = lines
+        .iter()
+        .map(|line| utils::get_display_length(line))
+        .collect();
+
+    let lines_display_width = match lines_display_width {
+        Ok(lines_display_width) => lines_display_width,
+        Err(e) => return Err(UpdateNotifierError::DisplayLengthError(e)),
+    };
+
+    // find the longest line to determine layout
+    let longest_line = lines_display_width
+        .iter()
+        .max()
+        .copied()
+        .unwrap_or_default();
+    let full_message_width = longest_line + DEFAULT_PADDING;
+
+    // create a curried render function to reduce verbosity when calling
+    let render_at_layout = |layout: utils::Layout, width: usize| {
+        utils::render_message(
+            layout,
+            width,
+            lines,
+            lines_display_width,
+            full_message_width,
+        )
+    };
+
+    // render differently depending on viewport
+    if let Some((Width(term_width), _)) = size {
+        // if possible, pad this value slightly
+        let term_width = if term_width > 2 {
```

Comment:
Just wondering, what are the units here? It's not characters, right?

Reply:
It is chars, or "columns"